### DISCUSSION TITLE
BO that support synchronous and ansynchronous factory methods

### DISCUSSION THREAD

**Discussion by bashinamwila** [MarimerLLC/csla#1817] [Questions] [answered]
I want BO's to support both synchronous and asynchronous static factory methods.
But if I include both versions of the Data Portal methods(one for sync and one for async) I am getting a Data Portal Exception.
For example
`[Create]
public void Create(){}` and
`[Create]
public async Task CreateAsync(){}`
Many thanks

**Comment by rockfordlhotka** [comment, marked as answer]
I think you misunderstand how the data portal works, as I don't think there is any reason to do what you are describing.

The _client-side_ data portal is where you _invoke_ the data portal:

```c#
  var obj1 = DataPortal.Create<MyBusinessType>();
  var obj2 = await DataPortal.CreateAsync<MyBusinessType>();
```

The _server-side_ data portal is where you implement data portal operation methods. Whether a data portal operation is implemented as an async method depends entirely on whether your method needs to invoke an async API to create/fetch/update the data.

```c#
  [Create]
  private void Create()
  {
    // create the object without calling any async APIs
  }
```

or

```c#
  [Create]
  private Task CreateAsync()
  {
    // create the object by calling async APIs
  }
```

It doesn't matter whether your _client-side_ calls are sync or async, that depends on how you want your UI to work.

The data portal takes _either type_ of client-side call, relays it to the app server, and invokes the server-side data portal operation method.
